snuba/consumers/consumer_builder.py

```diff
@@ -62,6 +65,19 @@ class MockParameters:
     std_deviation: int
 
 
+def validate_passed_slice(
+    storage_key: StorageKey, slice_id: Optional[int] = None
+) -> None:
+    """
+    Verifies that the given storage can be sliced
+    and that the slice_id passed in is within the range
+    of the total number of slices for the given storage
+    """
+    if slice_id is not None:
+        assert storage_key.value in SLICED_STORAGES
+        assert slice_id < SLICED_STORAGES[storage_key.value]
+
+
```

Comment:
Please do this inside the shared `KafkaTopicSpec` implementation as well. There is no reason to implement it multiple times inside every entrypoint.

Reply:
I'm not sure how this would be moved into `KafkaTopicSpec`. This validation has no relationship with any Topic/Topics as a concept. This is simply checking whether a slice_id (an int) and a corresponding storage_key are "compatible' (using those two assert statements). There is no reference to Topic maps or creation of sliced topics here. In order to put this into `KafkaTopicSpec`, we would need to pass storage_key into some function of `KafkaTopicSpec` -- does that make sense to you?

Reply:
Ok you are right. KafkaTopicSpec may not be the right place since it's not storage aware. But how do we prevent this code from being redefined in every endpoint? And ideally, how do we also ensure this function gets automatically called on any attempt to fetch a sliced storage instead of relying on the user to remember to do so?

Reply:
Good point. I will think about how I can make this part of an automatic validation and update.

Reply:
Let me ask from a different angle? What would happen if we passed a storage that was not actually sliced in here with a slice ID, and vice versa? Would it fail anyway and at which later point?

Reply:
Yes, it would fail when trying to retrieve the cluster. So, `build_batch_writer` in __build_streaming_strategy_factory calls `get_batch_writer` which calls `get_cluster` for some storage set key and slice id. If the storage and slice id pair are incompatible, this will fail and return an `UndefinedClickhouseCluster` (as you can see in the `get_cluster` function).

Reply:
Update: we've decided to move the validation step to `partitioning.py`, but this and underlying discussions (slicing by storage vs storage set) will be revisited.